### PR TITLE
Pelican style Markdown metadata (Issue #1923)

### PR DESCRIPTION
**Pull request by ralsina**
This uses the markdown "meta" extension to read metadata in a manner compatible with Pelican (Issue #1923)

It still presents a few problems:

* Performance: file is read and converted twice.
* ~~This doesn't really support translated sources, same problem is present in ipynb plugin, because it uses
  source_path instead of translated_source_path. OTOH, translated_source_path doesn't work yet at this stage :-P~~

### CODE REVIEW

File: nikola/plugins/compile/markdown/__init__.py [hunk 1]
```diff
@@ -134,3 +141,17 @@ def create_post(self, path, **kw):
...
+    def read_metadata(self, post, file_metadata_regexp=None, unslugify_titles=False, lang=None):
+        """Read the metadata from a post, and return a metadata dict."""
+        if not self.support_metadata:
+            return {}
+        if Markdown is None:
+            req_missing(['markdown'], 'build this site (compile Markdown)')
+        if lang is None:
+            lang = LocaleBorg().current_lang
+        # FIXME: what about translations????
+        source = post.source_path
```
Comment (Kwpolska) [inline nikola/plugins/compile/markdown/__init__.py]: `post.translated_source_path(lang)`

Also, note that Pelican uses Title-Cased metadata fields, and has some different names.

Comment (ralsina) [inline nikola/plugins/compile/markdown/__init__.py]: translated_source_path sort of crashes, I need to debug that. Code in ipynb plugin has the same problem.

I'll probably lower-case the keys then.

**Comment by ralsina** [comment]
From invariance results, it seems this is taking priority over actual explicit metadata (at least for title) which is not good, I think.

File: nikola/post.py [hunk 2]
```diff
@@ -153,15 +153,17 @@ def __init__(
...
+
         default_metadata, self.newstylemeta = get_meta(self, self.config['FILE_METADATA_REGEXP'], self.config['UNSLUGIFY_TITLES'])
 
         self.meta = Functionary(lambda: None, self.default_lang)
         self.meta[self.default_lang] = default_metadata
 
-        # Load internationalized metadata
         for lang in self.translations:
-            if os.path.isfile(get_translation_candidate(self.config, self.source_path, lang)):
-                self.translated_to.add(lang)
```
Comment (Kwpolska) [inline nikola/post.py]: Perhaps we should add a fallback to `get_translation_candiate` if `self.translated_to` is empty? There also seems to be a `lang != self.default_lang` case that will return the source path (shouldn’t it be flipped?)

File: nikola/plugins/compile/rest/__init__.py [hunk 3]
```diff
@@ -59,6 +59,38 @@ class CompileRest(PageCompiler):
...
     logger = None
 
+    def read_metadata(self, post, file_metadata_regexp=None, unslugify_titles=False, lang=None):
+        """Read the metadata from a post, and return a metadata dict."""
+        if lang is None:
+            lang = LocaleBorg().current_lang
+        source = post.translated_source_path(lang)
+
+        with io.open(source, 'r', encoding='utf-8') as inf:
+            document = docutils.core.publish_doctree(
```
Comment (Kwpolska) [inline nikola/plugins/compile/rest/__init__.py]: I’m not a reST expert, but perhaps there is a way to make this reusable and store it for `compile_string`?

Also, we might need to remove field lists from the output, if metadata is imported.

File: .gitignore [hunk 4]
```diff
@@ -195,5 +195,5 @@ Session.vim
 *~
 
 # Demo for Bagguette
```
Comment (Kwpolska) [inline .gitignore]: this should be at the top, before the gitignore.io template starts

**Comment by Kwpolska** [comment]
```diff
-<meta property="article:published_time" content="2012-03-30T23:00:00-03:00">
+<meta property="article:published_time" content="2013-12-31T23:59:59Z">
```

Somehow, the `.. date:` from Nikola comment metadata is ignored and replaced by the invariant date ([over here](https://github.com/getnikola/nikola/blob/2623322c6315b0fa30e9dd1b78e64ddf2e3bb14a/nikola/post.py#L200)).

**Comment by Kwpolska** [comment]
Sites using `post-list` crash:

```pytb
  File "/Users/kwpolska/git/nikola/nikola/plugins/compile/rest/post_list.py", line 181, in run
    filename = self.state.document.settings._nikola_source_path
AttributeError: 'Values' object has no attribute '_nikola_source_path'
```

Possible fixes: set that attribute in the metadata thing; disable directives in the metadata thing; make that line failsafe. (@ralsina, you decide)

File: nikola/plugins/compile/rest/__init__.py (hunk 3 above)
Comment (ralsina) [inline nikola/plugins/compile/rest/__init__.py]: Yeah to both things. We can probably keep the doctree somewhere in the Post object and reuse it if it's there.

Same about markdown, actually.

File: nikola/plugins/compile/rest/__init__.py [hunk 5]
```diff
@@ -59,6 +59,38 @@ class CompileRest(PageCompiler):
     demote_headers = True
     logger = None
 
+    def read_metadata(self, post, file_metadata_regexp=None, unslugify_titles=False, lang=None):
+        """Read the metadata from a post, and return a metadata dict."""
+        if lang is None:
+            lang = LocaleBorg().current_lang
+        source = post.translated_source_path(lang)
+
+        with io.open(source, 'r', encoding='utf-8') as inf:
```
Comment (ralsina) [inline nikola/plugins/compile/rest/__init__.py]: Good point.

File: nikola/post.py (hunk 2 above)
Comment (ralsina) [inline nikola/post.py]: Not really sure. I'll try to take a harder look.

File: .gitignore (hunk 4 above)
Comment (ralsina) [inline .gitignore]: Ok, changing it.

File: .gitignore [hunk 6]
```diff
@@ -14,6 +17,8 @@ output/
 
 # All of .idea should be ignored
 .idea/
+.vscode/
```
Comment (ralsina) [inline .gitignore]: And I really like it! except the git "support" which is garbage.

**Pull request merged**